### PR TITLE
Detect Shadow DOM Element using POM

### PR DESCRIPTION
**Pull request by manas-droid**
1. Shadow DOM is detected with the help of Page Object Model.
2. Self healing and Auto Update for shadow DOM

Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
  - Introduced classes and methods for detecting and interacting with web elements using the Page Object Model (POM) in Selenium WebDriver, including handling Shadow DOM elements.
  - Added JavaScript functionality to retrieve all shadow roots in a document for enhanced element detection.

- **Refactor**
  - Updated constructors and method parameters for improved integration and functionality.
  - Enhanced element testing logic to support various platform types.

- **Chores**
  - Adjusted and added import statements in multiple files for necessary libraries and dependencies.

- **Documentation**
  - Made minor comment formatting adjustments in the code for improved clarity.

- **Bug Fixes**
  - Resolved an issue related to XPath list initialization for shadow element location.

### CODE REVIEW

File: Ginger/Ginger/UserControlsLib/UCElementDetails.xaml.cs [hunk 1]
```diff
@@ -1051,9 +1053,20 @@ private void TestElementButtonClicked(object sender, RoutedEventArgs e)
...
+                    var htmlElementInfo = new HTMLElementInfo() { Path = testElement.Path, Locators = testElement.Locators, Properties = SelectedElement.Properties};
+                    WindowExplorerDriver.TestElementLocators(htmlElementInfo);
                 }
 
-                WindowExplorerDriver.TestElementLocators(testElement);
+                else
+                {
+                    WindowExplorerDriver.TestElementLocators(testElement);
+
+                }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/UserControlsLib/UCElementDetails.xaml.cs:1069]: ok, will implement this

File: Ginger/Ginger/UserControlsLib/UCElementDetails.xaml.cs [hunk 2]
```diff
@@ -1069,6 +1082,11 @@
         {
             if (xActUIPageFrame.Content != null && xActUIPageFrame.Content is ControlActionsPage_New)
             {
+                if((xActUIPageFrame.Content as ControlActionsPage_New).DefaultAction is ActUIElement && SelectedElement is HTMLElementInfo htmlElementInfo)
+                {
+                    ((xActUIPageFrame.Content as ControlActionsPage_New).DefaultAction as ActUIElement).XPathListToLocateShadowElement = htmlElementInfo.XPathList;
+                }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/UserControlsLib/UCElementDetails.xaml.cs:1088]: XPathList in the HTMLElementInfo class is never null. It has a default List as its initial value

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SearchElementsFromDOM.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            {
+                ReadOnlyCollection<object> childNodes = (ReadOnlyCollection<object>)((IJavaScriptExecutor)driver).ExecuteScript("return arguments[0].childNodes", searchContext);
+                return childNodes.Where((childNode) => childNode is IWebElement).Select((childNode)=>(IWebElement)childNode).ToList();
+            }
+
+            catch(InvalidCastException)
+            {
+                ReadOnlyCollection<IWebElement> childNodes = (ReadOnlyCollection<IWebElement>)((IJavaScriptExecutor)driver).ExecuteScript("return arguments[0].childNodes", searchContext);
+                return childNodes.ToList();
+            }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SearchElementsFromDOM.cs:46]: InvalidCastException in seleniumDriver to get childNodes is thrown when the List is not of the expected Generic type defined in the "ReadOnlyCollection". Therefore, the generic is either "object" or "IWebElement".

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SearchElementsFromDOM.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+            foreach (IWebElement child in childNodes)
+            {
+                try
+                {
+                    webElement = shadowDOM.FindShadowRootDirectChild(shadowRoot, locator, seleniumDriver.mDriver, child.TagName) ?? locateWebElement.LocateElementByLocator(locator, child, null, false);
+                }
+                catch
+                {
+                    continue;
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SearchElementsFromDOM.cs:126]: ok, as there is a try and catch used inside the functions inside the functions being called in the loop. I'll remove this try and catch

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/DetectDOMElements.cs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        DetectDOMElements was created to Detect Web Elements through Page Object Model
+    */
+    public class DetectDOMElements
+    {
+        SeleniumDriver seleniumDriver;
+        ShadowDOM shadowDOM = new();
+        public DetectDOMElements(SeleniumDriver seleniumDriver)
+        {
+            this.seleniumDriver = seleniumDriver;
+        }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/DetectDOMElements.cs:24]: ok , will implement this

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SearchElementsFromDOM.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SearchElementsFromDOM.cs:46]: ok, Thanks will implement this!

File: Ginger/GingerCoreNET/RunLib/Common/HTMLElementInfo.cs [hunk 6]
```diff
@@ -119,5 +123,21 @@ public override string GetAbsoluteXpath()
...
 
+
+        public HTMLElementInfo FindParentElementUsingGuid(IList<ElementInfo> AllElements)
+        {
+            if (AllElements == null)
+            {
+                return null;
+            }
+            var ParentPOMGuid = FindParentPOMGuid();
+            return (HTMLElementInfo)AllElements.FirstOrDefault((element) => element.Guid.Equals(new Guid(ParentPOMGuid)));
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/RunLib/Common/HTMLElementInfo.cs:134]: Handle empty guid condition

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/ApplicationModelsLib/POMModels/POMWizardLib/DeltaWizard/PomDeltaViewPage.xaml.cs [hunk 7]
```diff
@@ -431,7 +432,24 @@ private void HighlightElementClicked(object sender, RoutedEventArgs e)
 
             if (mSelectedElement != null)
             {
-                mWinExplorer.HighLightElement(mSelectedElement.ElementInfo, true);
+                if(mWinExplorer is SeleniumDriver selenium) 
+                {
+                    IList<ElementInfo> elementInfos = new ObservableList<ElementInfo>();
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/ApplicationModelsLib/POMModels/POMWizardLib/DeltaWizard/PomDeltaViewPage.xaml.cs:437]: ok, I have implemented this

File: Ginger/Ginger/ApplicationModelsLib/POMModels/POMWizardLib/DeltaWizard/PomDeltaViewPage.xaml.cs [hunk 8]
```diff
@@ -104,12 +107,21 @@ DeltaElementLocator mSelectedLocator
...
         }
 
-        public PomDeltaViewPage(ObservableList<DeltaElementInfo> deltaElements = null, GridColView gridCompareColumn = null, Agent mAgent = null)
+        private void SetElementsInfoList()
+        {
+            foreach(var element in mDeltaElements.Select(Delta=>Delta.ElementInfo))
+            {
+                elementInfoList.Add(element);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/ApplicationModelsLib/POMModels/POMWizardLib/DeltaWizard/PomDeltaViewPage.xaml.cs:116]: This function is only called once and when before it's called the deltaElements List is initialized.

File: Ginger/Ginger/ApplicationModelsLib/POMModels/POMWizardLib/DeltaWizard/PomDeltaViewPage.xaml.cs [hunk 9]
```diff
@@ -457,8 +472,13 @@ private void TestElementButtonClicked(object sender, RoutedEventArgs e)
             if (mSelectedLocator != null)
             {
                 bool originalActiveVal = mSelectedLocator.ElementLocator.Active;
-                mSelectedLocator.ElementLocator.Active = true;//so it will be tested even if disabeled 
-                mWinExplorer.TestElementLocators(new ElementInfo() { Path = CurrentEI.ElementInfo.Path, Locators = new ObservableList<ElementLocator>() { mSelectedLocator.ElementLocator } });
+                mSelectedLocator.ElementLocator.Active = true;//so it will be tested even if disabeled
+                var ElementInfo = CurrentEI.ElementInfo;
+                ElementInfo.Locators = [mSelectedLocator.ElementLocator];
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/ApplicationModelsLib/POMModels/POMWizardLib/DeltaWizard/PomDeltaViewPage.xaml.cs:477]: This is a new way to initialize List according to .Net  8

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs [hunk 10]
```diff
@@ -3922,8 +3922,9 @@ private void SwitchFrame(ElementInfo EI)
                 }
             }
         }
-
-        public IWebElement LocateElementByLocators(ElementInfo currentPOMElementInfo, bool iscallfromFriendlyLocator = false, POMExecutionUtils POMExecutionUtils = null)
+        
+        // TODO: convert recursion to iteration    
+        public IWebElement LocateElementByLocators(ElementInfo currentPOMElementInfo, IList<ElementInfo> MappedUIElements, bool iscallfromFriendlyLocator = false, POMExecutionUtils POMExecutionUtils = null)
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Selenium/SeleniumDriver.cs:3927]: Can you show how can the recursive code be converted to iterative for this function

**Review by Maheshkale447 — APPROVED** [review]

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**